File: 1. credit/credit.c
#include <cs50.h>
#include <math.h>
#include <stdbool.h>
#include <stdio.h>

int count_digits(long long n);
bool is_valid_length(int count);
bool luhn_algorithm(long long card_number);
void print_card_type(long long card_number);

int main(void)
{
    long long card_number;
    do
    {
        card_number = get_long_long("Number: ");
    }
    while (card_number <= 0);

    if (luhn_algorithm(card_number))
    {
        print_card_type(card_number);
    }
    else
    {
        printf("INVALID\n");
    }

    return 0;
}

int count_digits(long long n)
{
    int count = 0;
    while (n != 0)
    {
        n /= 10;
        count++;
    }
    return count;
}

bool is_valid_length(int count)
{
    return (count == 13 || count == 15 || count == 16);
}

bool luhn_algorithm(long long card_number)
{
    int sum = 0;
    bool multiply = false;
    while (card_number > 0)
    {
        int digit = card_number % 10;
        if (multiply)
        {
            digit *= 2;
            while (digit > 0)
            {
                sum += digit % 10;
                digit /= 10;
            }
        }
        else
        {
            sum += digit;
        }
        card_number /= 10;
        multiply = !multiply;
    }
    return (sum % 10 == 0);
}
void print_card_type(long long card_number)
{
    int count = count_digits(card_number);
    if (is_valid_length(count))
    {
        int first_two_digits = card_number / (long long) (pow(10, count - 2));
        if ((count == 15 && (first_two_digits == 34 || first_two_digits == 37)))
        {
            printf("AMEX\n");
        }
        else if (count == 16 && (first_two_digits >= 51 && first_two_digits <= 55))
        {
            printf("MASTERCARD\n");
        }
        else if ((count == 13 || count == 16) && (first_two_digits / 10 == 4))
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}

File: 6. sentimental/credit/credit.py
from cs50 import get_string

def main():
    # Prompt user for input
    card_number = get_string("Number: ")

    # Validate input
    if not card_number.isdigit() or not (13 <= len(card_number) <= 16):
        print("INVALID")
        return

    # Check if the card number is valid using Luhn's algorithm
    if not luhn_check(card_number):
        print("INVALID")
        return

    # Determine card type
    card_type = get_card_type(card_number)

    # Print result
    print(card_type)

def luhn_check(card_number):
    # Reverse the card number
    card_number = card_number[::-1]

    # Initialize variables
    total = 0

    # Iterate over every other digit, starting with the second-to-last digit
    for i in range(1, len(card_number), 2):
        digit = int(card_number[i])
        # Double the digit and add the digits of the product
        total += digit * 2 % 10 + digit * 2 // 10

    # Add the sum of the digits that were not doubled
    total += sum(int(card_number[i]) for i in range(0, len(card_number), 2))

    # Check if the total modulo 10 is zero
    return total % 10 == 0

def get_card_type(card_number):
    # Determine card type based on the starting digits and length
    if card_number[:2] in ["34", "37"] and len(card_number) == 15:
        return "AMEX"
    elif card_number[:2] in [str(i) for i in range(51, 56)] and len(card_number) == 16:
        return "MASTERCARD"
    elif card_number[0] == "4" and (len(card_number) == 13 or len(card_number) == 16):
        return "VISA"
    else:
        return "INVALID"

if __name__ == "__main__":
    main()
